### ISSUE TITLE
Linux and MacOS Support

### ISSUE THREAD

**Issue by Wolfmyths** [Wolfmyths/Genshin-Stopwatch#8]
I want Genshin Stopwatch to be able to run on all desktop platforms. However, I do not know how to change `.py` files into an executable for Linux and MacOS systems. If someone can comment down below on how to achieve this that'd be great.

Genshin Stopwatch is running off of the PyQt5 framework which is cross platform and should work be able to work on MacOS and Linux. But I cannot test to see if Genshin Stopwatch works on either platforms since I only own a windows machine.

~~Issue #2 needs to be fixed before this issue can close because it uses a module exclusive to windows called `win10toast`~~

**Comment by H4ppy-04** [comment]
> I want Genshin Stopwatch to be able to run on all desktop platforms. However, I do not know how to change .py files into an executable for Linux and MacOS systems. If someone can comment down below on how to achieve this that'd be great.

(TLDR read option 3; it's probably the one you want)

**Option 1**
By adding `#!/usr/bin/env python` to the top of `main.py`. It gives information to the operating system. In the context of a Linux distribution, it stipulates what program should be used to execute the file.

Then, by making the file executable with `chmod +x main.py`, the program can be run as an 'executable' (note the parenthesis; not run as a literal executable).
```sh
./main
```

> Users can then run the file assuming they have navigated to the file location

This approach requires navigating to the program location and manually executing the file. 
This may not always be what you want. For example, if you'd like the program name to be different, you can create a custom installer, which is something I'm doing for a [project](https://github.com/GH-Syn/urban-cli) I'm working on. This leads on nicely to option 2.

**Option 2**
Option 2 uses option 1, except instead of the user manually navigating to a specific directory and running `./main`, an installation `.sh` script can be created. The shell script adds the project to a global path. This is a fairly simple process, and can be directly replicated from [uban-cli](https://github.com/GH-Syn/urban-cli/blob/main/install.sh).

**Option 3**
Option 3 is the easiest, as it can be compiled beforehand. This option involves pre-compiling the executable and including the file in the [release](https://docs.github.com/en/repositories/releasing-projects-on-github). released as an executable, removing the need for users manually compiling the program.

If this option were to be implemented, it would most likely be integrated as an [action](https://docs.github.com/en/actions/learn-github-actions/understanding-github-actions#overview) that would install and run [pyinstaller](https://pyinstaller.org/en/stable/usage.html).

**Comment by Wolfmyths** [comment]
Yes option 3 does seem to be the best option. When I make a new release, I use `pyinstaller` myself to compile the script into an `.exe` so the end user doesn't have to. The release file doesn't include the source code. Can actions automate this process?

**Comment by H4ppy-04** [comment]
> When I make a new release, I use `pyinstaller` myself to compile the script into an `.exe` so the end-user doesn't have to. The release file doesn't include the source code. Can actions automate this process?

Yep :)

You can also _automate_ releases! So for example, if you wanted to only release and compile a build when a new tag has been created, that is quite possible.

If this were to be made, a separate branch would have to be made for testing. Luckily, I've forked the repository so testing can be done without interfering with the development of this repository.

**Comment by H4ppy-04** [comment]
**A quick update**
 - I've had a crack at trying to get the installer working on Linux. 

 - Unfortunately, the win10toast dependency fails to install on a Linux distribution due to the nature of the library. 

**Screenshots**

<img height="350px" src="https://github.com/Wolfmyths/Genshin-Stopwatch/assets/101031214/2c690df9-f17e-4e58-9e7f-d43a1b668cb2" alt="demo-save">

> The installation runs and produces an executable

<br>

<img width="350px" src="https://github.com/Wolfmyths/Genshin-Stopwatch/assets/101031214/4404b98c-92e2-42c1-bcef-15fb61c50be6" alt="exception-occurs">

 **(1)** Add win10toast to `requirements.txt`
   > installing win10toast via pip doesn't work on Linux

 **(2)** Fails when executing due to unmet requirements, suggesting that requirements need installation before compiling into an executable.

Note: requirements.txt exists to run `python -m pip install --user -r requirements.txt` so that dependencies don't have to be installed individually

**Comment by Wolfmyths** [comment]
Yeah I was afraid that would happen. That's why I want issue #2 to be completed first so we can have a way to use notifications on all desktop platforms. In issue #2 I have details on the module I want to use.

I would add in this myself but I can't figure out a way to thread the notification function using the module. `win10toast` does this automatically through a bool parameter (See line 733 & 754 in [main.py](https://github.com/Wolfmyths/Genshin-Stopwatch/blob/styleSheet-Rewrite/main.py#L735))

**Comment by Wolfmyths** [comment]
Linux & MacOS should work now because of [this commit](https://github.com/Wolfmyths/Genshin-Stopwatch/commit/baf2ee9a04e32d763f16af0c998de904eac81e1c) but I'm unable to test it myself

**Comment by H4ppy-04** [comment]
> Linux & MacOS should work now because of [this commit](https://github.com/Wolfmyths/Genshin-Stopwatch/commit/baf2ee9a04e32d763f16af0c998de904eac81e1c) but I'm unable to test it myself

Testing it now

**Edit** it runs! :partying_face:

**Comment by Wolfmyths** [comment]
That's great to hear! Could you maybe test the notification function too? Just set a custom timer to 1 minute and see if it fires off

**Comment by H4ppy-04** [comment]
The notification daemon doesn't seem to work :disappointed: 
With that said, it is pretty hard to get notifications to work on linux because linux will have all types of different notification managers and some of them may not work well with certain libraries.

**Comment by Wolfmyths** [comment]
Yeah I heard that might be an issue. Gonna have to use github actions to test if ubuntu works I guess? I'll also have to implement a try except block. The apprise documentation might have a fix, there were some other options for specifically linux distros

**Comment by Wolfmyths** [comment]
**Update on MacOS support:** I found someone to try running Genshin Stopwatch for me and it all works but there's a catch, MacOS users need something called `terminal-notifier`. There's [documentation](https://github.com/caronc/apprise/wiki/Notify_macosx) on how to install it. And then the user needs to allow permissions for `terminal-notifier`. I'm not sure if we can automate this process for the user but for now I think this is something the user should do manually as it's pretty easy to install.

Also I found a module called [py2app](https://pypi.org/project/py2app/) which just looks like pyinstaller but for MacOS users

Edit: **py2app can only be ran on MacOS** so we'll have to use some kind of virtual environment to test this
